src/storage/src/source/mysql/snapshot.rs

```diff
@@ -0,0 +1,281 @@
+use futures::TryStreamExt;
+use mysql_async::prelude::Queryable;
+use mysql_async::{IsolationLevel, Row as MySqlRow, TxOpts};
+use mz_sql_parser::ast::display::AstDisplay;
+use timely::dataflow::channels::pact::Pipeline;
+use timely::dataflow::operators::{Broadcast, CapabilitySet, Concat, ConnectLoop, Feedback, Map};
+use timely::dataflow::{Scope, Stream};
+use timely::progress::{Antichain, Timestamp};
+use tracing::trace;
+
+use mz_mysql_util::{
+    ensure_full_row_binlog_format, ensure_gtid_consistency, ensure_replication_commit_order,
+    query_sys_var, GtidSet, SchemaRequest,
+};
+use mz_mysql_util::{schema_info, MySqlTableDesc};
+use mz_repr::{Diff, GlobalId, Row};
+use mz_sql_parser::ast::UnresolvedItemName;
+use mz_storage_types::sources::MySqlSourceConnection;
+use mz_timely_util::builder_async::{
+    Event as AsyncEvent, OperatorBuilder as AsyncOperatorBuilder, PressOnDropButton,
+};
+
+use crate::source::{RawSourceCreationConfig, SourceReaderError};
+
+use super::{pack_mysql_row, ReplicationError, RewindRequest, TransactionId, TransientError};
+
+/// Renders the snapshot dataflow. See the module documentation for more information.
+pub(crate) fn render<G: Scope<Timestamp = TransactionId>>(
+    mut scope: G,
+    config: RawSourceCreationConfig,
+    connection: MySqlSourceConnection,
+    subsource_resume_uppers: BTreeMap<GlobalId, Antichain<TransactionId>>,
+    table_info: BTreeMap<UnresolvedItemName, (usize, MySqlTableDesc)>,
+) -> (
+    Collection<G, (usize, Result<Row, SourceReaderError>), Diff>,
+    Stream<G, RewindRequest>,
+    Stream<G, ReplicationError>,
+    PressOnDropButton,
+) {
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("MySqlSnapshot({})", config.id), scope.clone());
+
+    let (mut raw_handle, raw_data) = builder.new_output();
+    let (mut rewinds_handle, rewinds) = builder.new_output();
+    let (mut _definite_error_handle, definite_errors) = builder.new_output();
+
+    // Broadcast a signal from the snapshot leader worker to the other workers when the table
+    // lock is in place. Upon receiving the first message the workers should start a transaction
+    // that guarantees they all see a consistent view from the same GTID.
+    let (mut lock_start_handle, lock_start) = builder.new_output();
+    let (ls_feedback_handle, ls_feedback_data) = scope.feedback(Default::default());
+    let mut lock_start_input = builder.new_disconnected_input(&ls_feedback_data, Pipeline);
+    lock_start.broadcast().connect_loop(ls_feedback_handle);
+
+    // Broadcast a signal from the all workers that they have begun a transaction and that the
+    // table lock held by the snapshot leader can be released
+    let (mut transaction_start_handle, transaction_start) = builder.new_output();
+    let (ts_feedback_handle, ts_feedback_data) = scope.feedback(Default::default());
+    let mut transaction_start_input = builder.new_disconnected_input(&ts_feedback_data, Pipeline);
```

Comment:
I think this pact should be `Exchange(|_| snapshot_leader_id)` so that receives all the acknowledgements that a transaction has started at some worker